php/php.editor/src/org/netbeans/modules/php/editor/csl/FoldingScanner.java

```diff
@@ -179,7 +179,7 @@ public Map<String, List<OffsetRange>> folds(ParserResult info) {
             program.accept(new FoldingVisitor(folds));
             Source source = phpParseResult.getSnapshot().getSource();
             assert source != null : "source was null";
-            Document doc = source.getDocument(false);
+            Document doc = source.getDocument(true);
```

Comment:
@junichi11 would be good to observe this to check if it causes any performance issues.

It could be worth checking how often `source.getDocument(false)` would be null here during regular project development. If its very often the case, we could try to solve this differently (e.g bullet point 1 of the PR)

Reply:
@mbien 
Or, change to `true` only when we test?

```diff
diff --git a/php/php.editor/src/org/netbeans/modules/php/editor/csl/FoldingScanner.java b/php/php.editor/src/org/netbeans/modules/php/editor/csl/FoldingScanner.java
index fb565eb64a..730455feb3 100644
--- a/php/php.editor/src/org/netbeans/modules/php/editor/csl/FoldingScanner.java
+++ b/php/php.editor/src/org/netbeans/modules/php/editor/csl/FoldingScanner.java
@@ -141,6 +141,7 @@ public final class FoldingScanner {
 
     private static final String LAST_CORRECT_FOLDING_PROPERTY = "LAST_CORRECT_FOLDING_PROPERY"; //NOI18N
     private static final boolean FOLD_PHPTAG = !Boolean.getBoolean("nb.php.editor.doNotFoldPhptag"); // NOI18N NETBEANS-5480
+    static volatile boolean IS_UNIT_TEST = false;
 
     public static FoldingScanner create() {
         return new FoldingScanner();
@@ -179,7 +180,8 @@ public final class FoldingScanner {
             program.accept(new FoldingVisitor(folds));
             Source source = phpParseResult.getSnapshot().getSource();
             assert source != null : "source was null";
-            Document doc = source.getDocument(true);
+            // avoid getting `null` in unit testing
+            Document doc = source.getDocument(IS_UNIT_TEST);
             if (FOLD_PHPTAG) {
                 processPHPTags(folds, doc);
             }
diff --git a/php/php.editor/test/unit/src/org/netbeans/modules/php/editor/csl/FoldingTest.java b/php/php.editor/test/unit/src/org/netbeans/modules/php/editor/csl/FoldingTest.java
index 083d201589..dec5a55fb1 100644
--- a/php/php.editor/test/unit/src/org/netbeans/modules/php/editor/csl/FoldingTest.java
+++ b/php/php.editor/test/unit/src/org/netbeans/modules/php/editor/csl/FoldingTest.java
@@ -30,6 +30,12 @@ public class FoldingTest extends PHPTestBase {
         super(testName);
     }
 
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        FoldingScanner.IS_UNIT_TEST = true;
+    }
+
     public void testFoldingMethod() throws Exception {
         checkFolds("testfiles/parser/foldingMethod.php");
     }

```

Reply:
yes but making the program aware of the test is always a bit ugly. The program should, if possible, never know that it is tested so that the results are the same as in real world. This is also a valid option but I would try to use it only if we really have to.

Reply:
OK. Perhaps, someone would report performance issues if there are them :) Thanks!

Reply:
but something weird is happening here, we might have to revert this:

both micronaut and LSP jobs are failing after this merge and I can't get them green again. I can reproduce it locally, if i set back to false, `MicronautGradleArtifactsImplTest` is green, true makes it red...

Reply:
Umm, I set back to `false` but unit tests fail...

```
Product Version: Apache NetBeans IDE 19
Java: 19.0.2; OpenJDK 64-Bit Server VM 19.0.2+7-Ubuntu-0ubuntu322.04
Runtime: OpenJDK Runtime Environment 19.0.2+7-Ubuntu-0ubuntu322.04
System: Linux version 6.2.0-36-generic running on amd64; UTF-8;
```
![nb-micrronaut-test](https://github.com/apache/netbeans/assets/738383/e4578039-bd34-416d-be8c-91cb38311042)

Reply:
yes, there is something else going on, I try to debug it here: https://github.com/mbien/netbeans/pull/1

Reply:
(this has likely nothing to do with this PR or any recent PRs, both micronaut and LSP tests which fail use gradle which is suspicious, the runner images, jdk version, everything remained unchanged since the last successful build on master)

Reply:
https://github.com/apache/netbeans/pull/6673

Reply:
Great. Thanks!